[css-color] Fix serialization of percentages in color-mix()

nt1m opened the pull request:
#### fdf9b00e784ec3133bed99073957900ae9a03381
<pre>
[css-color] Fix serialization of percentages in color-mix()
<a href="https://bugs.webkit.org/show_bug.cgi?id=259829">https://bugs.webkit.org/show_bug.cgi?id=259829</a>
rdar://113399146

Reviewed by Darin Adler.

From <a href="https://drafts.csswg.org/css-color-5/#serial-color-mix">https://drafts.csswg.org/css-color-5/#serial-color-mix</a>

&gt; The serialization of the specified value of a color-mix() function is the string &quot;color-mix(in &quot;, followed by the specified &lt;color-space&gt; in all-lowercase,
&gt; followed by &quot;, &quot;, followed by the first specified color, followed by a space, followed by the specified (un-normalized) first percentage (unless both percentages are 50%),
&gt; followed by &quot;, &quot;, followed by the second specified color, followed by the specified (un-normalized) second percentage (unless the two specified percentages add to 100%), followed by &quot;)&quot;.

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-mix-function.html:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/StyleColor.cpp:
(WebCore::serializationForCSS):
* Source/WebCore/css/color/CSSColorMixSerialization.h: Added.
(WebCore::serializeColorMixColor):
(WebCore::serializeColorMixPercentage):
(WebCore::percentageDoubleValue):
(WebCore::serializationForCSSColorMix):
* Source/WebCore/css/color/CSSUnresolvedColorMix.cpp:
(WebCore::serializationForCSS):

Canonical link: <a href="https://commits.webkit.org/267596@main">https://commits.webkit.org/267596@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6253e656d78e982273926d21fd7a1a002b07e14c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17106 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17431 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17920 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18892 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16006 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17298 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20705 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17573 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18211 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17309 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17665 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14849 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19708 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14905 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15540 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22232 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15904 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15707 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20049 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16292 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/13821 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15449 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/15380 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4085 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19814 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16125 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/color/CSSUnresolvedColorMix.h:
```diff
@@ -55,6 +67,39 @@ struct CSSUnresolvedColorMix {
 String serializationForCSS(const CSSUnresolvedColorMix&);
 
+// https://drafts.csswg.org/css-color-5/#serial-color-mix
+template<typename T>
```
I would have named the template argument `ColorMixType`.

darinadler commented on Source/WebCore/css/StyleColor.h:
```diff
@@ -146,6 +146,18 @@ struct StyleColorMix {
     struct Component {
         StyleColor color;
         std::optional<double> percentage;
+        void serializeColor(StringBuilder& builder)
```
Seems to me these helper functions to make the template function work don't have to be member functions. They could just be free functions that are overloaded for both color mix types; to me they have more to do with adapting the template to the two rather than with the intrinsic behavior of the two color mix objects themselves. If we did that they wouldn't have to be in this header. I worry about making the `StyleColor.h` header having more and more code in it. Not great for build times. And serialization only done in a few specific places.

nt1m committed:
[css-color] Fix serialization of percentages in color-mix()
https://bugs.webkit.org/show_bug.cgi?id=259829
rdar://113399146

Reviewed by Darin Adler.

From https://drafts.csswg.org/css-color-5/#serial-color-mix

> The serialization of the specified value of a color-mix() function is the string "color-mix(in ", followed by the specified <color-space> in all-lowercase,
> followed by ", ", followed by the first specified color, followed by a space, followed by the specified (un-normalized) first percentage (unless both percentages are 50%),
> followed by ", ", followed by the second specified color, followed by the specified (un-normalized) second percentage (unless the two specified percentages add to 100%), followed by ")".

* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-mix-function-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-color/parsing/color-valid-color-mix-function.html:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/StyleColor.cpp:
(WebCore::serializationForCSS):
* Source/WebCore/css/color/CSSColorMixSerialization.h: Added.
(WebCore::serializeColorMixColor):
(WebCore::serializeColorMixPercentage):
(WebCore::percentageDoubleValue):
(WebCore::serializationForCSSColorMix):
* Source/WebCore/css/color/CSSUnresolvedColorMix.cpp:
(WebCore::serializationForCSS):

Canonical link: https://commits.webkit.org/267596@main
